### PR TITLE
ENH: Add Convolutional Neural Network Models

### PR DESCRIPTION
**Pull request by richford**
This PR is a resubmission of #72 after fixing some style issues and using `git filter-repo` to remove some checkpoint and weights files from the history.

### CODE REVIEW

**Comment by richford** [comment]
Phew, ignore the massive number of commits above. I rebased and now there are only 13 commits.

File: afqinsight/cnn.py [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+        model.load_weights(weights_path)
+
+        # return the model
+        return model
+
+    def build_basic_model(self, X, y):
+        """Build a sequential model without hyperparameter tuning.
+
+        Builds a static, basic sequential model with no hyperparameter tuning with the
+        architecture used to produce state of the art Weston Havens results.
```
Comment (arokem) [inline afqinsight/cnn.py:237]:
```suggestion
        Builds a static baseline sequential model with no hyperparameter tuning.
```

Comment (arokem) [reply afqinsight/cnn.py:237]: Maybe we can add the WH comment in a `Notes` section below? Or use this in an example (probably a separate PR).

File: afqinsight/cnn.py [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+        strategy="median",
+        random_state=200,
+        directory=None,
+        project_name=None,
+        **tuner_kwargs,
+    ):
+        """Construct a CNN.
+
+        Use the given number of n_nodes, each with a max depth of max_depth.
+        """
```
Comment (arokem) [inline afqinsight/cnn.py:337]: We probably want a little bit more documentation of arguments/kwargs here.

File: afqinsight/cnn.py [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+        self.project_name = project_name
+        self.tuner_kwargs = tuner_kwargs
+        self.model_ = None
+        self.best_hps_ = None
+
+    def _preprocess(self, X, y=None):
+        """Convert feature matrix for input into a CNN.
+
+        Masks NAN values for X and y (if y is given), imputes X, and reshapes X
+        to be in proper form for CNN model. In more conventionall machine
```
Comment (arokem) [inline afqinsight/cnn.py:405]:
```suggestion
        to be in proper form for CNN model. In more conventional machine
```

File: afqinsight/cnn.py [hunk 4]
```diff
@@ -0,0 +1,548 @@
...
+        learning, X has shape (n_samples, n_features), where n_features is
+        n_nodes * n_bundles * n_metrics. However, in our CNN approach, we treat
+        each bundle/metric combination as a separate channel, analogous to RGB
+        channels in a 2D image. The remaining one dimension is the nodes
+        dimension. Thus the output has shape (n_samples, n_channels, n_nodes),
+        where n_channels = n_metrics * n_bundles.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
```
Comment (arokem) [inline afqinsight/cnn.py:415]: I think we want this docstring to be a bit more clear. This is the output of afqi.load_data, so I think that this should be fine:

```suggestion
        X : array-like of shape (n_samples, n_metrics * n_nodes)
```

File: afqinsight/cnn.py [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+                The feature samples
```
Comment (arokem) [inline afqinsight/cnn.py:416]:
```suggestion
                Diffusion MRI tractometry features (columns) for each subject in the sample (rows).
```

File: afqinsight/cnn.py [hunk 6]
```diff
@@ -0,0 +1,548 @@
...
+    def fit(self, X, y):
+        """Fit the model.
+
+        Preprocesses X and y, builds CNN model, tunes model hyperparameters and
+        fits the model to given X and y, using X_test and y_test to validate and
+        find best weights and hyperparameters.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
```
Comment (arokem) [inline afqinsight/cnn.py:468]: Here as well: 

```suggestion
        X : array-like of shape (n_samples, n_metrics * n_nodes)
```

File: afqinsight/cnn.py [hunk 7]
```diff
@@ -0,0 +1,548 @@
...
+                The feature samples
```
Comment (arokem) [inline afqinsight/cnn.py:469]:
```suggestion
                Diffusion MRI tractometry features (columns) for each subject (rows).
```

File: afqinsight/cnn.py [hunk 8]
```diff
@@ -0,0 +1,548 @@
...
+
+    def predict(self, X):
+        """Predict target values.
+
+        Preprocesses X and returns predicted y values for X from fitted CNN model.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+                The feature samples
```
Comment (arokem) [inline afqinsight/cnn.py:515]:
```suggestion
        X : array-like of shape (n_samples, n_metrics * n_nodes)
                Tractometry features (columns) for each subject in the sample (rows).
```

File: afqinsight/cnn.py [hunk 9]
```diff
@@ -0,0 +1,548 @@
...
+        """
+        X = self._preprocess(X)
+        check_is_fitted(self, "is_fitted_")
+        pred = self.model_.predict(X).squeeze()
+        return pred
+
+    def score(self, y_test, y_hat):
+        """Score the performance of the model.
+
+        Applies a NAN mask to y_test and returns r-squared score for the CNN model.
```
Comment (arokem) [inline afqinsight/cnn.py:530]:
```suggestion
        Masks out NaN values from y_test and returns $R^2$ score for the CNN model comparing to y_hat
```

**Review by arokem — COMMENTED** [review]
Sorry that it has taken me so long to review. I think that this is basically ready to merge, unless someone has any objections). All of my comments here are about documentation, and except for one comment asking to specify the parameters to the constructor of the CNN object, none are really crucial, in my opinion. 

@richford : do you have any time to work on this, or do you want me to make a PR with these things, so that you can then merge it and I can merge this? Or would you rather wait a bit longer with this? I don't think there's any urgency to merge this.

**Comment by arokem** [comment]
I believe that once #85 is merged this will be ready to go.

**Pull request merged**